[_]: feature/use-multipart-uploads-on-backups

patricioxavier8 opened the pull request:
In this PR the following changes were made:
Added retries when a part fails.
Use part sizes of 50MB
If the file is >100MB, use multipart uploads
Use taskgroup to upload up to 6 parts at the same time
Use actors to safely handle concurrency

patricioxavier8 committed:
change enum uploadError to public and add new error when part upload fail

patricioxavier8 committed:
- change multipart min size to 200
- change multipart chunk size to 50
- add retries when part upload fails

patricioxavier8 committed:
add equatable in upload error to fix test cases

patricioxavier8 committed:
-add ProgressHandlerStore to manage concurrency
- sort shardPayload

patricioxavier8 committed:
-add UploadedPartsActor and UploadState to manage concurrency
- add TaskGroup and DispatchSemaphore to upload 6 parts at the same time

sg-gs commented on Sources/InternxtSwiftCore/Services/Network/NetworkFacade.swift:
```diff
@@ -129,63 +128,132 @@ public struct NetworkFacade {
+                    }
+                    let uploadedPartConfig = UploadedPartConfig(etag: etag, partNumber: partIndex + 1)
+                    await uploadedPartsActor.addUploadedPartConfig(uploadedPartConfig)
+                    return
```
This `return` is not needed

patricioxavier8 replied:
The return is necessary here to prevent the method from re-entering the loop if the upload is successful.

sg-gs commented on Sources/InternxtSwiftCore/Services/Network/NetworkFacade.swift:
```diff
@@ -129,63 +128,132 @@ public struct NetworkFacade {
+        
+        let uploadedPartsActor = UploadedPartsActor()
+        let uploadState = UploadState()
+        let semaphore = DispatchSemaphore(value: 6)
```
Your explanation is on the right track, but it could benefit from some clarification and slight rewording for better readability and technical precision. Here’s a refined version:

Using [OperationQueue](https://developer.apple.com/documentation/foundation/operationqueue) is a better approach because it allows you to enqueue new tasks (e.g., the next chunks of uploads) dynamically. This ensures that the concurrency limit is utilized efficiently, even if some tasks take longer to complete.

For example, imagine you have 6 uploads running concurrently, with the following durations:

- 1st upload: 5s
- 2nd upload: 5s
- 3rd upload: 10s
- 4th upload: 15s
- 5th upload: 5s
- 6th upload: 5s

In this case, the 4th upload, which takes 15 seconds, would block the entire upload from processing additional part uploads until it completes. If most uploads take only 5 seconds, a slow upload (like the 4th) would leave resources idle, as other uploads have already finished.

By using an `OperationQueue`, new uploads (e.g., the 7th, 8th, etc.) can start immediately when previous tasks complete, filling in the available slots and keeping the system fully utilized. This effectively masks the delays caused by slower uploads, ensuring optimal performance.

In contrast, a DispatchSemaphore does not manage task completion dynamically and would result in underutilized concurrency in scenarios like this.

patricioxavier8 committed:
- add async operation to use operation queue
-add network monitor to check user connection

patricioxavier8 committed:
- add operationqueue to upload 6 parts at same time
- Implemented a mechanism to wait for network connectivity during multipart upload.
- Added a timeout (`MAX_WAIT_TIME`) to abort the upload if the connection is not restored within the specified duration.
- Ensures the upload process handles network interruptions gracefully.
- Uses `Task.sleep` to check connectivity at 20-second intervals.

sg-gs commented on Sources/InternxtSwiftCore/Services/Network/NetworkFacade.swift:
```diff
@@ -114,78 +115,117 @@ public struct NetworkFacade {
-            // and the stream reading is stopped
-            try await processEncryptedChunk(encryptedChunk: encryptedChunk, partIndex: partIndex)
-            partIndex += 1
+            tasks.append(encryptedChunk)
```
It's easier and cleaner to do this way, but a high RAM usage issue will appear, proportional to the file size. Using this way to enqueue chunks means that to start the upload, you need the same RAM size as the file size, which breaks the scalability of multiparts (and limits the max file size to upload to the entire RAM size, at best).

An efficient implementation would be to enqueue each chunk as soon as you get it and wait for the queue to have a free operation to enqueue the next chunk. For instance, assume 6 operations at the same time, in pseudo-code: 

```
fileReader = getFileReader()

while queue.isNotBusy():
    queue.enqueue(fileReader.getNextChunk())
```

Where `isNotBusy()` is a function that returns true as soon as the queue has less than 6 active tasks ongoing.An equivalence in Swift would be to use a semaphore inside the upload function like: 

```
let operation = BlockOperation {
    let semaphore = DispatchSemaphore(value: 0)
    uploadChunk(nextChunk) {
        semaphore.signal()
    }
    semaphore.wait() // Wait until upload completes
}

operation.completionBlock = {
    enqueueNextChunk() // Enqueue the next chunk when this one finishes
}
```

_This is just a suggestion_, but I hope you get the idea. Therefore, you fill just enough RAM to have 6-7 chunks in memory (50*6=300MB) instead of the entire size of the file while avoiding long-lived HTTP requests which tend to fail more often.

patricioxavier8 committed:
Fix memory usage during large file uploads by releasing encrypted chunks after each operation completes and optimizing OperationQueue management

sg-gs commented on Sources/InternxtSwiftCore/Services/Network/NetworkFacade.swift:
```diff
@@ -132,40 +132,56 @@ public struct NetworkFacade {
-            
-            let uploadUrl = uploadUrls[index]
+
+            if await uploadedPartsActor.isPartUploaded(partIndex: partIndex) {
```
If this function is being called for each encrypted chunk, and as far as I see, it is not parallelized, why is this required? We should not aim to skip any chunk but wait to enqueue the next as you do a bit below correctly.

sg-gs commented on Sources/InternxtSwiftCore/Services/Network/NetworkFacade.swift:
```diff
@@ -310,7 +326,7 @@ public struct NetworkFacade {
     
     class UploadPartOperation: AsyncOperation {
-        let encryptedChunk: Data
+        var encryptedChunk: Data?
```
Why does this need to be changed to `var`?

Let's avoid overwriting a variable's value to handle binary data as that could lead to unintended consequences like corruption of the data, data loss, etc.

patricioxavier8 committed:
fix upload progress

patricioxavier8 committed:
remove unnecesary validation
